lib/procbot.ts

```diff
@@ -0,0 +1,265 @@
+
+/**
+ * A map linking contexts to Worker instances.
+ * @typedef {Object} ProcBot.WorkerMap
+ * @property {generic} - A generic context for the Worker.
+ * @property {Worker}
+ */
+type WorkerMap<T> = Map<T, Worker<T>>;
+
+/**
+ * The Worker class is responsible for the execution of scheduling tasks based on events.
+ * @class ProcBot.Worker
+ * @classdesc
+ * Each Worker instance is bound to a context. This context could be, for example, a
+ * unique Github repository, or a directory in a file system, a specific customer service
+ * in a set, etc. It is also generic, and can therefore be of any type.
+ * When WorkerEvents are added to an empty queue, they are processed for being worked on
+ * in the next tick of event loop.
+ */
+export class Worker<T> {
+    /**
+     * Holds the context for the Worker.
+     * @member Worker._context
+     * @private
+     * @type {generic}
+     */
+    private _context: T;
+    /**
+     * Reference to the parent Map in which this Worker exists.
+     * @member Worker.parentMap
+     * @private
+     * @type {generic}
+     */
+    private parentMap: WorkerMap<T>;
+    /**
+     * Holds the queue of events to work on.
+     * @member Worker.queue
+     * @private
+     * @type {ProcBot.WorkerEvent[]}
+     */
+    private queue: WorkerEvent[] = [];
+
+    /**
+     * Creates the Worker class, specifying a context and the parent Map.
+     * @param {generic} context - The context to use for hashing.
+     * @param {ProcBot.WorkerMap} parentMap - The parent Map containing all Workers.
+     */
+    constructor(context: T, parentMap: WorkerMap<T>) {
+        this._context = context;
+        this.parentMap = parentMap;
+    }
+
+    /**
+     * Retrieve the context for the Worker.
+     * @return {generic} - The context.
+     */
+    get context() {
+        return this._context;
+    }
+
+    /**
+     * Add a new event to the Worker's event queue.
+     * @param {ProcBot.WorkerEvent} worker - The event to add to the queue.
+     */
+    addEvent(event: WorkerEvent) {
+        this.queue.push(event);
+        // If this is a new worker, ensure it operates.
+        if (this.queue.length === 1) {
+            this.runWorker();
+        }
+    }
+
+    // Run as many workers as are queued. Do this atomically, in FIFO order.
+    /**
+     * Runs the next worker in the queue, before deleting its entry. Should more entries
+     * exist it then runs those.
+     */
+    private runWorker() {
+        // Get the next thing from the queue.
+        const entry = <WorkerEvent>this.queue.shift();
+        const self: this = this;
+
+        // Run worker, proceed to next worker.
+        entry.workerMethod(entry.event, entry.data)
+        .then(() => {
+            if (this.queue.length > 0) {
+                process.nextTick(this.runWorker);
+            } else {
+                // Unlink ourselves from our parent list.
+                self.parentMap.delete(self.context);
+            }
+        });
+    }
+}
```

Comment:
OK, this is where my C++ism really kicked in. TS doesn't let you (AFAIK, having read around) employ nested classes, which is how I'd have tacked this originally, with parts of the class as protected and all others private.

My issue here is that whilst some of the Worker types are to be exported, some (such as WorkerMap for example), I'd rather keep module private. If I split this up into a separate Worker file, I'd have to expose everything, right? Is there a good paradigm for doing this that isn't class inheritance, for example? (And hopefully that all makes sense!)